core/instance.go

```diff
@@ -291,36 +296,46 @@ func (i *instance) getCheapestCompatibleSpotInstanceType(allowedList []string, d
 	usedMappings := i.asg.launchConfiguration.countLaunchConfigEphemeralVolumes()
 	attachedVolumesNumber := min(usedMappings, current.instanceStoreDeviceCount)
 
+	// Find all compatible and not blocked instance types
 	for _, candidate := range i.region.instanceTypeInformation {
 
-		logger.Println("Comparing ", candidate.instanceType, " with ",
-			current.instanceType)
-
 		candidatePrice := i.calculatePrice(candidate)
+		logger.Println("Comparing candidate ", candidate.instanceType, " with price ", candidatePrice, " to current type ",
+			current.instanceType, " with price ", i.price)
 
-		if i.isPriceCompatible(candidatePrice, bestPrice) &&
+		if i.isPriceCompatible(candidatePrice) &&
 			i.isEBSCompatible(candidate) &&
 			i.isClassCompatible(candidate) &&
 			i.isStorageCompatible(candidate, attachedVolumesNumber) &&
 			i.isVirtualizationCompatible(candidate.virtualizationTypes) &&
 			i.isAllowed(candidate.instanceType, allowedList, disallowedList) {
-			bestPrice = candidatePrice
-			chosenSpotType = candidate.instanceType
-			cheapest = candidate
-			debug.Println("Best option is now: ", chosenSpotType, " at ", bestPrice)
-		} else if chosenSpotType != "" {
-			debug.Println("Current best option: ", chosenSpotType, " at ", bestPrice)
+			acceptableInstanceTypes = append(acceptableInstanceTypes, acceptableInstance{candidate, candidatePrice})
+			debug.Println("Found compatible and allowed instancetype: ", candidate.instanceType, " at ", candidatePrice, " added to launchcandiatelist")
+		} else if candidate.instanceType != "" {
+			debug.Println("Non compatible option found: ", candidate.instanceType, " at ", candidatePrice, " - discarding")
+		} else {
+
 		}
 	}
-	if chosenSpotType != "" {
-		debug.Println("Cheapest compatible spot instance found: ", chosenSpotType)
-		return cheapest, nil
+
+	if acceptableInstanceTypes != nil {
+		sort.Slice(acceptableInstanceTypes, func(i, j int) bool {
+			return acceptableInstanceTypes[i].price < acceptableInstanceTypes[j].price
+		})
+		debug.Println("List of cheapest compatible spot instances found, sorted ascending by price: ",
+			acceptableInstanceTypes)
+		var result []instanceTypeInformation
+		for _, ai := range acceptableInstanceTypes {
+			result = append(result, ai.instanceTI)
+		}
+		return result, nil
 	}
-	return cheapest, fmt.Errorf("No cheaper spot instance types could be found")
+
+	return nil, fmt.Errorf("No cheaper spot instance types could be found")
 }
 
 func (i *instance) launchSpotReplacement() error {
```

Comment:
The test coverage decreased because this function is not covered and it's now a bit bigger than before. 

I'd like to have this covered as well so that we're not decreasing our test coverage.